### PR TITLE
Add a rust module

### PR DESCRIPTION
**Pull request by dcbaker**
This is now ready for review.

The goal here is to add  a place to put some helpers to make rust and meson play nice together. Presumably this would be the place to put cargo integration into meson as well.

What I have right now is a single helper for building rust tests, it takes an existing rust build target, and creates a new executable build target with the same sources and arguments, but adds the `--test` argument, as well as the `--format=pretty` option, and sets the test protocol to `'rust'`. This makes turning on rust unit tests super simple, as the updated test case shows.

I've chosen not to add this as unstable because we're going to start using this in mesa as soon as it lands.

### CODE REVIEW

**Comment by dcbaker** [comment]
Ah, also note that I used a bunch of python 3.6 stuff in this, so some test failures on platforms still using 3.5 are expected. Since 3.5 support is being dropped for 0.57, I went ahead and used nice, new stuff.

**Comment by dcbaker** [comment]
Anyone interested in looking at this?

**Comment by nirbheek** [comment]
I am really glad that you're working on this! I want to review this, but I am not sure when I will be able to get to it. 

> I've chosen not to add this as unstable because we're going to start using this in mesa as soon as it lands.

Is this because you want the API to be fixed before you start using it in Mesa or you're worried about breakage from the rename? `import('unstable-rust')` will continue to work even after we remove the prefix and call the module `rust`, so that shouldn't be a problem.

However, if you want the API to be finalized before merging, that would make things harder. Often actual usage shows that an API break is needed. But maybe that's ok, and we can just wrap it into our usual deprecation cycle, like we did for the Python module.

**Comment by dcbaker** [comment]
Ah, I didn't realize the `unstable-rust` would continue to work if it was renamed. I can change it then, as that was my biggest concern.

**Comment by dcbaker** [comment]
I'll change to `unstable-rust` and I think I'll split some of the groundwork patches out as they should be safe to merge separately.

**Comment by bonzini** [comment]
Regarding "_mtest: Handle subtest results as a dict_",  the pull request at #8029 is already adding a lot of the infrastructure to store TAP test results as individual objects (e.g. `TAPParser.Test` objects).

**Comment by dcbaker** [comment]
And if 8029 lands first I'll rebase on top of it. This has been opened for quite a while (and the rust test type changes which included the dict change were opened a long time before that).

**Comment by jpakkane** [comment]
And that one is waiting for this one. Which one do people want to see landed first?

**Comment by bonzini** [comment]
I'm fine with this one going in first if @dcbaker rebases it.

**Comment by dcbaker** [comment]
@jpakkane, @bonzini, rebased and passing CI (minus cygwin which still hasn't been fixed).

File: docs/markdown/Rust-module.md [hunk 1]
```diff
@@ -0,0 +1,32 @@
+---
+short-description: Rust language integration module
+authors:
+    - name: Dylan Baker
+      email: dylan@pnwbakers.com
+      years: [2020]
+...
+
+# Rust module
```
Comment (jpakkane) [inline docs/markdown/Rust-module.md:9]: This should mention that it is `unstable-rust` rather than `rust`, much like `unstable-simd`.

**Comment by jpakkane** [comment]
The module needs a release note snippet.

**Pull request merged**